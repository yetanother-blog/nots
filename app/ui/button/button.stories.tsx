import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

const meta: Meta<typeof Button> = {
  title: 'Forms/Button',
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    children: 'Primary Button',
  },
};
