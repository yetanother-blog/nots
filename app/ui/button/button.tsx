import React from "react";
import { clsx } from "clsx/lite";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {}

export const Button: React.FC<ButtonProps> = ({
  children,
  className,
  ...props
}) => {
  const classes = clsx("rounded bg-blue-500 text-white px-4 py-3", className);

  return (
    <button className={classes} {...props}>
      {children}
    </button>
  );
};
