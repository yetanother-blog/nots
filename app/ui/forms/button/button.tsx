import { forwardRef } from 'react';
import { clsx } from 'clsx';
import { Typography } from '~/ui/data-display/typography/typography';
import type { IconName } from '~/ui';
import { LoadingIndicator, Icon } from '~/ui';
import type { PolymorphicComponentPropWithRef } from '~/ui/utils/polymorphic-types';

type Variant = 'filled';
export type ButtonColor = 'grey';
type Size = 'regular';

export type ButtonProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<
    C,
    {
      isLoading?: boolean;
      variant?: Variant;
      color?: ButtonColor;
      size?: Size;
      iconPosition?: 'start' | 'end';
      icon?: IconName;
      className?: string;
    }
  >;

type ButtonComponent = <C extends React.ElementType = 'button'>(
  props: ButtonProps<C>
) => React.ReactElement | null;

// eslint-disable-next-line react/display-name
export const Button = forwardRef(
  <C extends React.ElementType = 'button'>(
    {
      as,
      children,
      isLoading,
      disabled,
      color = 'grey',
      size = 'regular',
      variant = 'filled',
      icon,
      iconPosition = 'end',
      className,
      ...props
    }: ButtonProps<C>,
    ref?: unknown
  ) => {
    /*
     * Filled Variants
     */
    const filledClasses = clsx([
      'text-white dark:text-nots-grey-800',
      isLoading &&
        'drop-shadow-none hover:drop-shadow-none pointer-events-none',
    ]);

    const primaryFilledClasses = clsx([
      filledClasses,
      'disabled:bg-nots-grey-200 dark:disabled:bg-white dark:disabled:text-nots-grey-200 focus:outline-none',
      'bg-nots-grey-800 dark:bg-white hover:bg-nots-grey-600 dark:hover:bg-nots-grey-100 focus:outline-none focus:ring-2 focus:ring-nots-grey-300',
      isLoading && 'hover:bg-ws-primary-500',
    ]);

    const colorClasses = clsx([
      variant === 'filled' && color === 'grey' && primaryFilledClasses,
    ]);

    const baseShapeClasses = clsx([size === 'regular' && `px-3 py-[6px]`]);

    const iconOnlyClasses = clsx(['w-9 h-8', 'py-0', 'px-0']);
    const isOnlyIcon = !children && icon;

    const classes = clsx([
      'flex items-center justify-center rounded-sm',
      'transition-colors duration-200 ease-in-out',
      'gap-2',
      isOnlyIcon ? iconOnlyClasses : baseShapeClasses,
      iconPosition === 'start' && 'flex-row-reverse',
      iconPosition === 'end' && 'flex-row',
      isLoading && 'pointer-events-none',
      colorClasses,
      className,
    ]);

    const hideClasses = clsx([isLoading && 'opacity-0 pointer-events-none']);

    const Component = as || 'button';

    return (
      <Component className={classes} disabled={disabled} {...props} ref={ref}>
        {children && (
          <Typography
            className={hideClasses}
            variant="heading3"
            fontWeight="semibold"
            as="span"
          >
            {children}
          </Typography>
        )}
        {isLoading && <LoadingIndicator className="absolute" />}
        {icon && <Icon icon={icon} className={hideClasses} />}
      </Component>
    );
  }
) as ButtonComponent;
