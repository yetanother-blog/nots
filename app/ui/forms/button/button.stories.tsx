import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

/**
 * The Button component is used to trigger an action or event, such as
 * submitting a form, opening a dialog, canceling an action, or performing a
 * delete operation.
 *
 * # Render as a Remix Router Link
 *
 * It's also possible to use the Button in combination with Remix Router (aka
 * React Router). Just pass the component of the routing library to the UI
 * component.
 *
 * ```tsx
 * import { Link } from '@remix-run/react';
 *
 * ...
 *
 * <Button as={Link} to="/some-route">
 *  Remix Router Link
 * </Button>
 * ```
 */
const meta: Meta<typeof Button> = {
  title: 'Forms/Button',
  component: Button,
  argTypes: {
    color: {
      control: {
        type: 'select',
        options: ['grey'],
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const GreyFilled: Story = {
  args: {
    color: 'grey',
  },
  render: () => (
    <div className="flex flex-grow-0 flex-wrap gap-4">
      <Button onClick={() => console.log('clicked')}>Primary Button</Button>
      <Button isLoading={true} onClick={() => console.log('clicked')}>
        Primary Button
      </Button>
      <Button disabled={true} onClick={() => console.log('clicked')}>
        Primary Button
      </Button>
      <Button
        onClick={() => console.log('clicked')}
        icon="caretDown"
        iconPosition="start"
      >
        Primary Button
      </Button>
      <Button onClick={() => console.log('clicked')} icon="caretDown">
        Primary Button
      </Button>
      <Button
        isLoading={true}
        onClick={() => console.log('clicked')}
        icon="caretDown"
      >
        Primary Button
      </Button>
      <Button
        disabled={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      >
        Primary Button
      </Button>
      <Button onClick={() => console.log('clicked')} icon="paragraph" />
      <Button
        isLoading={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      />
      <Button
        disabled={true}
        onClick={() => console.log('clicked')}
        icon="paragraph"
      />
    </div>
  ),
};
