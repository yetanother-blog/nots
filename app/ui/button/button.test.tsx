import { render } from "@testing-library/react";
import { Button } from "./button";

describe("primary button", () => {
  it("should match snapshot", () => {
    const { container } = render(<Button>Primary</Button>);
    expect(container).toMatchSnapshot();
  });
});
