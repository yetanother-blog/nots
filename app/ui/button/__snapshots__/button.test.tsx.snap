// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`primary button > should match snapshot 1`] = `
<div>
  <button
    class="rounded bg-blue-500 text-white px-4 py-3"
  >
    Primary
  </button>
</div>
`;
